Python/Product/Debugger/Debugger/DebugConnection.cs

```diff
@@ -0,0 +1,302 @@
+
+            try {
+                try {
+                    Debug.Assert(_connection != null);
+                    if (_connection != null) {
+                        lock (_isListeningLock) {
+                            _isListening = true;
+                            _listeningReadyEvent.Set();
+                        }
+                        await _connection.ProcessMessages();
+                    }
+                } catch (IOException) {
+                } catch (ObjectDisposedException ex) {
+                    // Socket or stream have been disposed
+                    Debug.Assert(
+                        ex.ObjectName == typeof(NetworkStream).FullName ||
+                        ex.ObjectName == typeof(Socket).FullName,
+                        "Accidentally handled ObjectDisposedException(" + ex.ObjectName + ")"
+                    );
+                } finally {
+                    lock (_isListeningLock) {
+                        // Exit out of the event handling thread
+                        _isListening = false;
+                        _eventsPendingWakeUp.Set();
+                    }
+                }
+
+                ProcessingMessagesEnded?.Invoke(this, EventArgs.Empty);
+            } catch (ThreadAbortException) {
+            }
+
+            Debug.WriteLine("MessageProcessingThreadAsync Ended");
+        }
+
+        private void EventHandlingThread() {
+            Debug.WriteLine("EventHandlingThread Started");
+            try {
+                _listeningReadyEvent.WaitOne();
+
+                while (true) {
+                    bool paused;
+                    lock (_isListeningLock) {
+                        if (!_isListening) {
+                            break;
+                        }
+
+                        paused = _isPaused;
+                    }
+
+                    EventReceivedEventArgs eventReceived = null;
+                    if (!paused && _eventsPending.TryDequeue(out eventReceived)) {
+                        try {
+                            HandleEvent(eventReceived);
+                        } catch (Exception e) when (!e.IsCriticalException()) {
+                            Debug.Fail(string.Format("Error while handling debugger event '{0}'.\n{1}", eventReceived.Name, e));
+                        }
+                    } else {
+                        _eventsPendingWakeUp.WaitOne();
+                    }
+                }
+            } catch (ThreadAbortException) {
+            }
+
+            Debug.WriteLine("EventHandlingThread Ended");
+        }
+
+        private void HandleEvent(EventReceivedEventArgs e) {
+            Debug.WriteLine(string.Format("PythonProcess handling event: {0}", e.Event.name));
+            lock (_eventHandlingLock) {
+                Debug.Assert(e.Event.name == LDP.LocalConnectedEvent.Name || e.Event.name == LDP.RemoteConnectedEvent.Name || _isAuthenticated);
+                switch (e.Event.name) {
```

Comment:
This will be shorter and typesafer if you use the new typeswitch on the event class:
```cs
switch (e.Event) {
case LDP.AsyncBreakEvent evt:
  LegacyAsyncBreak?.Invoke(this, evt);
  break;
case LDP.BreakpointFailedEvent evt:
  ...
}

```